wpengine/faustjs
# MERL-484: telemetry

## Description

This PR introduces a `faust` CLI that acts as a pass through to the `next` CLI, however, with our own telemetry that runs prior to the `next` CLI.

When running the `faust` CLI for the first time, you are asked if you would like to enroll in Faust Telemetry. By responding with either `true` or `false`, your preferences are then saved to your local machine (`~/.config/configstore/faust.json` on Mac) as a JSON file:

```jsonc
{
  telemetry: {
    notifiedAt: '12345', // timestamp the user was first notified about Faust telemetry
    anonymousId: '0000-0000-0000', // An anonymous v4 UUID for GA
    enabled: true // Either true or false based on user preferences
  }
}
```

From there, any time the `faust` CLI is ran, we check the users' preferences, and if telemetry is enabled, we send a GA event with relevant information.

In addition to the basic Next CLI commands like `faust start`, `faust dev`, and `faust build`, there is also a new command `faust faust-telemetry`, which allows the user to modify their telemetry preferences.

## Testing

1. Checkout PR branch
2. Uncomment this line: https://github.com/wpengine/faustjs/blob/61bee88bb152e00c206de939aa092f7558bd5313/packages/faustwp-cli/index.ts#L90
3. Run `npm run build -w @faustwp/core` from the monorepo root
4. Run `npm run build -w @faustwp/cli` from the monorepo root
5. Run `npm run dev -w @faustwp/getting-started-example` to start the Faust example app in dev mode
6. Notice the Telemetry prompt
7. Type `true`
8. You should now see a telemetry event being `console.log`'d
11. Now, run `npx faustnx faustnx-telemetry` and change your preferences to `false`
12. Run `npm run build -w faust-nx-getting-started` notice that the telemetry event is not being `console.log`'d

## Telemetry Data

The following data is what is beign sent to GA.

```ts
export interface TelemetryData {
  setting_has_frontend_uri?: boolean;
  setting_redirects_enabled?: boolean;
  setting_rewrites_enabled?: boolean;
  setting_themes_disabled?: boolean;
  setting_img_src_replacement_enabled?: boolean;
  faustwp_version?: string;
  is_wpe?: boolean;
  multisite?: boolean;
  php_version?: string;
  wp_version?: string;
  node_faustwp_core_version?: string;
  node_faustwp_cli_version?: string;
  node_apollo_client_version?: string;
  node_version?: string;
  node_next_version?: string;
  node_is_development?: boolean;
}
```
